File: app/src/main/java/com/streamvault/tv/data/catalog/FilmParser.kt
package com.streamvault.tv.data.catalog

import com.streamvault.tv.data.model.Episode
import com.streamvault.tv.data.model.Season
import com.streamvault.tv.data.model.Series
import org.jsoup.Jsoup
import java.net.URI

/**
 * Filmpalast-style movie catalog / detail parser.
 * Movies are modeled as [Series] with mediaKind="movie" and one synthetic season/episode.
 */
object FilmParser {
    private val EP_RE = Regex("""\bS\d{1,2}E\d{1,3}\b""", RegexOption.IGNORE_CASE)
    private val STREAM_RE = Regex("""/stream/[^/?#]+""", RegexOption.IGNORE_CASE)

    data class Hoster(
        val provider: String,
        val name: String,
        val url: String,
        val score: Int,
        val language: String = "",
    )

    fun browsePaths(): List<String> = listOf("/movies/new", "/movies/top", "/")

    fun isMovieSite(baseUrl: String): Boolean {
        val h = baseUrl.lowercase()
        return h.contains("filmpalast") || h.contains("movie") || h.contains("film")
    }

    fun isEpisodeLike(title: String?, url: String?): Boolean =
        EP_RE.containsMatchIn(title.orEmpty()) || EP_RE.containsMatchIn(url.orEmpty())

    /** Detect audio language from Filmpalast detail HTML / release title / URL. */
    fun detectPageLanguage(html: String, title: String = "", pageUrl: String = ""): String {
        val release = Regex(
            """id=["']release_text["'][^>]*>(.*?)</""",
            setOf(RegexOption.IGNORE_CASE, RegexOption.DOT_MATCHES_ALL)
        ).find(html)?.groupValues?.get(1).orEmpty()
        val fromMeta = Regex(
            """itemprop=["']inLanguage["'][^>]*content=["']([^"']+)["']""",
            RegexOption.IGNORE_CASE
        ).find(html)?.groupValues?.get(1)
        val genreEn = html.contains("/search/genre/Englisch", ignoreCase = true) &&
            (title.contains("ENGLISH", true) || pageUrl.contains("-english", true))
        return StreamLanguage.detectFromText(
            pageUrl,
            title,
            release,
            fromMeta,
            if (genreEn) "english" else null,
            html.take(6000),
        ) ?: StreamLanguage.DE
    }

    /**
     * Suggest sibling Filmpalast URLs for the other language (DE↔EN).
     * Filmpalast keeps German and English as separate /stream/{slug} pages.
     */
    fun siblingLanguageUrls(pageUrl: String, currentLang: String): List<String> {
        val uri = runCatching { URI(pageUrl) }.getOrNull() ?: return emptyList()
        val path = uri.path.orEmpty()
        val m = Regex("""^(.*?/stream/)([^/?#]+)/?$""", RegexOption.IGNORE_CASE).find(path)
            ?: return emptyList()
        val prefix = m.groupValues[1]
        val slug = m.groupValues[2]
        val candidates = linkedSetOf<String>()
        val lang = StreamLanguage.normalize(currentLang)
        if (lang == StreamLanguage.EN) {
            val stripped = slug
                .replace(Regex("""-english$""", RegexOption.IGNORE_CASE), "")
                .replace(Regex("""-eng$""", RegexOption.IGNORE_CASE), "")
                .replace(Regex("""-ovo$""", RegexOption.IGNORE_CASE), "")
            if (stripped.isNotBlank() && stripped != slug) {
                candidates += uri.resolve("$prefix$stripped").toString().trimEnd('/')
            }
            // minions-monsters-english → minions-monster (common Filmpalast plural quirk)
            val singular = stripped.replace(Regex("""s$"""), "")
            if (singular.isNotBlank() && singular != stripped) {
                candidates += uri.resolve("$prefix$singular").toString().trimEnd('/')
            }
        } else {
            candidates += uri.resolve("$prefix$slug-english").toString().trimEnd('/')
            candidates += uri.resolve("$prefix${slug}s-english").toString().trimEnd('/')
            candidates += uri.resolve("$prefix$slug-eng").toString().trimEnd('/')
        }
        return candidates.filter { it != pageUrl.trimEnd('/') }
    }

    fun languageFromMovieHit(title: String, url: String): String =
        detectPageLanguage("", title, url)

    fun parseMovieList(
        html: String,
        baseUrl: String,
        moviesOnly: Boolean = true,
    ): List<Series> {
        val doc = Jsoup.parse(html, baseUrl)
        val out = mutableListOf<Series>()
        val seen = linkedSetOf<String>()

        fun addFrom(hrefRaw: String, titleRaw: String, posterRaw: String?) {
            val href = abs(baseUrl, hrefRaw)
            if (!STREAM_RE.containsMatchIn(href) || !seen.add(href)) return
            val title = cleanTitle(titleRaw)
            if (title.isBlank()) return
            if (moviesOnly && isEpisodeLike(title, href)) return
            val poster = posterRaw?.takeIf { it.isNotBlank() }?.let { abs(baseUrl, it) }
            out += Series(
                id = slugId(href, title),
                title = title,
                posterUrl = poster,
                backdropUrl = poster,
                detailPath = href,
                mediaKind = "movie",
            )
        }

        doc.select("article.liste, article.glowliste, article").forEach { art ->
            val a = art.selectFirst("h2 a[href*=/stream/]")
                ?: art.selectFirst("a[href*=/stream/]")
                ?: return@forEach
            val href = a.attr("abs:href").ifBlank { a.attr("href") }
            val title = a.attr("title").ifBlank { a.text() }
            val img = art.selectFirst("img.cover-opacity, img.cover2, a img[src*=/files/movies/]")
                ?: art.selectFirst("img[src*=/files/movies/]")
            val poster = img?.let { imageSrc(it) }
            addFrom(href, title, poster)
        }

        if (out.isEmpty()) {
            doc.select("a[href*=/stream/]").forEach { a ->
                val href = a.attr("abs:href").ifBlank { a.attr("href") }
                val title = a.attr("title").ifBlank { a.text() }
                if (title.length < 2) return@forEach
                addFrom(href, title, null)
            }
        }
        return out
    }

    fun parseMovieDetail(html: String, pageUrl: String, idHint: String? = null): Series {
        val doc = Jsoup.parse(html, pageUrl)
        val title = cleanTitle(
            doc.selectFirst("article.detail h2, h2.bgDark, h2")?.text()
                ?: doc.selectFirst("meta[itemprop=name], .fn .value-title")?.attr("title")
                ?: doc.title()
        )
        val overview = doc.selectFirst("[itemprop=description]")?.text()?.trim()
            ?: doc.selectFirst("cite > span.hidden")?.text()?.trim()
        val year = Regex("""Veröffentlicht:\s*(\d{4})""", RegexOption.IGNORE_CASE)
            .find(html)?.groupValues?.get(1)?.toIntOrNull()

        var poster: String? = null
        val posterEl = doc.selectFirst("img.cover2")
            ?: doc.selectFirst("img[itemprop=image], img[src*=/files/movies/]")
        if (posterEl != null) {
            poster = abs(pageUrl, imageSrc(posterEl).orEmpty()).takeIf { it.isNotBlank() }
        }
        if (poster.isNullOrBlank()) {
            Regex("""/files/movies/\d+/[^"'\\\s>]+\.(?:jpg|jpeg|png|webp)""", RegexOption.IGNORE_CASE)
                .find(html)?.value?.let { poster = abs(pageUrl, it) }
        }

        val pageLang = detectPageLanguage(html, title, pageUrl)
        val genres = doc.select("#detail-content-list a[href*=/search/genre/]")
            .map { cleanTitle(it.text()) }
            .filter { it.isNotBlank() }
            .distinct()
            .toMutableList()
        // Surface audio language as a genre chip for detail UI (Deutsch / Englisch).
        val langLabel = StreamLanguage.label(pageLang)
        if (genres.none { it.equals(langLabel, true) || it.equals("Deutsch", true) || it.equals("Englisch", true) }) {
            genres.add(0, langLabel)
        }

        val id = idHint?.takeIf { it.isNotBlank() } ?: slugId(pageUrl, title)
        return Series(
            id = id,
            title = title,
            overview = overview,
            year = year,
            posterUrl = poster,
            backdropUrl = poster,
            detailPath = pageUrl,
            mediaKind = "movie",
            genres = genres,
            availableLanguages = listOf(pageLang),
            languagePages = mapOf(pageLang to pageUrl),
            seasons = listOf(
                Season(
                    number = 1,
                    title = "Film",
                    posterUrl = poster,
                    backdropUrl = poster,
                    episodes = listOf(
                        Episode(
                            id = "$id-movie",
                            seriesId = id,
                            seasonNumber = 1,
                            number = 1,
                            title = title.ifBlank { "Film" },
                            overview = overview,
                            stillUrl = poster,
                            streamPageUrl = pageUrl,
                            streamUrl = null,
                        )
                    )
                )
            )
        )
    }

    fun scoreHoster(
        name: String,
        url: String = "",
        language: String = "",
        preferredLang: String = StreamLanguage.DE,
    ): Int {
        val n = "$name $url".lowercase()
        var s = 0
        // Firestream first for Filmpalast: progressive CDN, fewer geo/encoding fails than VOE.
        if (n.contains("firestream")) s += 120
        if (Regex("""\bvoe\b""").containsMatchIn(n) || n.contains("voe.sx")) s += 100
        if (n.contains("vidara") || n.contains("vidnest")) s += 70
        if (n.contains("vidsonic")) s += 40
        if (n.contains("playmate")) s += 20
        if (Regex("""\bhd\b""").containsMatchIn(n)) s += 5
        if (language.isNotBlank()) {
            if (StreamLanguage.matchesPreferred(language, preferredLang)) s += 80
            else s -= 40
        }
        return s
    }

    fun parseHosters(
        html: String,
        pageUrl: String,
        preferredLang: String = StreamLanguage.DE,
    ): List<Hoster> {
        val doc = Jsoup.parse(html, pageUrl)
        val pageLang = detectPageLanguage(
            html,
            doc.selectFirst("article.detail h2, h2.bgDark, h2")?.text().orEmpty(),
            pageUrl,
        )
        val hosters = mutableListOf<Hoster>()
        doc.select("ul.currentStreamLinks").forEach { ul ->
            val name = ul.selectFirst(".hostName")?.text()?.trim().orEmpty().ifBlank { "Hoster" }
            // Prefer live data-player-url; skip commented markup; accept plain iconPlay hrefs.
            val a = ul.select("a[data-player-url]").firstOrNull { it.attr("data-player-url").isNotBlank() }
                ?: ul.select("a.iconPlay[href], a.button.iconPlay[href], a.button[href]")
                    .firstOrNull { href ->
                        val h = href.attr("abs:href").ifBlank { href.attr("href") }
                        h.isNotBlank() && h != "#" && !h.contains("javascript:", true)
                    }
                ?: return@forEach
            val raw = a.attr("data-player-url").ifBlank { a.attr("abs:href") }.ifBlank { a.attr("href") }
            if (raw.isBlank() || raw == "#") return@forEach
            val url = abs(pageUrl, raw)
            if (url.isBlank()) return@forEach
            val hosterLang = a.attr("data-language-label").ifBlank { a.attr("data-language") }
                .ifBlank { ul.attr("data-language") }
                .ifBlank { pageLang }
            val score = scoreHoster(name, url, hosterLang, preferredLang)
            hosters += Hoster(
                provider = name,
                name = name,
                url = url,
                score = score,
                language = hosterLang,
            )
        }
        return hosters.sortedByDescending { it.score }
    }

    fun cleanTitle(raw: String): String =
        raw.replace(Regex("""\s+"""), " ")
            .replace(Regex("""\*\d{4}\*"""), "")
            .trim()

    fun slugId(url: String, title: String): String {
        val path = runCatching { URI(url).path }.getOrNull().orEmpty()
        val slug = Regex("""/stream/([^/?#]+)""", RegexOption.IGNORE_CASE)
            .find(path)?.groupValues?.get(1)
            ?: path.split('/').filter { it.isNotBlank() }.lastOrNull()
            ?: title.ifBlank { "movie" }
        return slug.lowercase().replace(Regex("[^a-z0-9]+"), "-").trim('-')
    }

    fun abs(base: String, href: String): String {
        if (href.isBlank()) return ""
        if (href.startsWith("//")) return "https:$href"
        if (href.startsWith("http://") || href.startsWith("https://")) return href
        return runCatching { URI(base).resolve(href).toString() }.getOrDefault(href)
    }

    private fun imageSrc(el: org.jsoup.nodes.Element): String? {
        val raw = el.attr("abs:src").ifBlank { el.attr("src") }
            .ifBlank { el.attr("abs:data-src") }
            .ifBlank { el.attr("data-src") }
            .ifBlank { el.attr("content") }
        return raw.takeIf { it.isNotBlank() && !it.startsWith("data:") }
    }
}

File: webapp/renderer/film-parser.js
/** Filmpalast-style movie catalog / detail parser (also used as movies-site template). */
window.FilmParser = (() => {
  const EP_RE = /\bS\d{1,2}E\d{1,3}\b/i;
  const STREAM_RE = /\/stream\/[^/?#]+/i;

  function abs(base, href) {
    if (!href) return "";
    try {
      if (href.startsWith("//")) return `https:${href}`;
      return new URL(href, base).toString();
    } catch {
      return href;
    }
  }

  function cleanTitle(raw) {
    return String(raw || "")
      .replace(/\s+/g, " ")
      .replace(/\*\d{4}\*/g, "")
      .trim();
  }

  function slugId(url, title) {
    let path = "";
    try {
      path = new URL(url).pathname;
    } catch {
      path = url;
    }
    const slug =
      (path.match(/\/stream\/([^/?#]+)/i) || [])[1] ||
      path.split("/").filter(Boolean).pop() ||
      title ||
      "movie";
    return String(slug)
      .toLowerCase()
      .replace(/[^a-z0-9]+/g, "-")
      .replace(/^-|-$/g, "");
  }

  function isEpisodeLike(title, url) {
    return EP_RE.test(title || "") || EP_RE.test(url || "");
  }

  function isMovieSite(baseUrl) {
    const h = String(baseUrl || "").toLowerCase();
    return /filmpalast|movie|film/.test(h);
  }

  /** Browse paths for movie catalogs on filmpalast-like sites. */
  function browsePaths() {
    return ["/movies/new", "/movies/top", "/"];
  }

  function parseMovieList(html, baseUrl, { moviesOnly = true } = {}) {
    const doc = new DOMParser().parseFromString(html, "text/html");
    const out = [];
    const seen = new Set();

    doc.querySelectorAll("article.liste, article.glowliste, article").forEach((art) => {
      const a =
        art.querySelector('h2 a[href*="/stream/"]') ||
        art.querySelector('a[href*="/stream/"]');
      if (!a) return;
      const href = abs(baseUrl, a.getAttribute("href") || "");
      if (!STREAM_RE.test(href) || seen.has(href)) return;
      const title = cleanTitle(a.getAttribute("title") || a.textContent);
      if (!title) return;
      if (moviesOnly && isEpisodeLike(title, href)) return;

      const img =
        art.querySelector("img.cover-opacity, img.cover2, a img[src*='/files/movies/']") ||
        art.querySelector("img[src*='/files/movies/']");
      const poster = img ? abs(baseUrl, img.getAttribute("src") || "") : null;
      const id = slugId(href, title);
      seen.add(href);
      out.push({
        id,
        title,
        posterUrl: poster,
        backdropUrl: poster,
        detailPath: href,
        mediaKind: "movie",
        overview: null,
        year: null,
        genres: [],
        seasons: [],
      });
    });

    // Fallback: naked stream anchors
    if (!out.length) {
      doc.querySelectorAll('a[href*="/stream/"]').forEach((a) => {
        const href = abs(baseUrl, a.getAttribute("href") || "");
        if (!STREAM_RE.test(href) || seen.has(href)) return;
        const title = cleanTitle(a.getAttribute("title") || a.textContent);
        if (!title || title.length < 2) return;
        if (moviesOnly && isEpisodeLike(title, href)) return;
        seen.add(href);
        out.push({
          id: slugId(href, title),
          title,
          posterUrl: null,
          backdropUrl: null,
          detailPath: href,
          mediaKind: "movie",
          overview: null,
          year: null,
          genres: [],
          seasons: [],
        });
      });
    }
    return out;
  }

  function parseMovieDetail(html, pageUrl, idHint) {
    const doc = new DOMParser().parseFromString(html, "text/html");
    const title =
      cleanTitle(doc.querySelector("article.detail h2, h2.bgDark, h2")?.textContent) ||
      cleanTitle(doc.querySelector('meta[itemprop="name"], .fn .value-title')?.getAttribute("title")) ||
      cleanTitle(doc.title);
    const overview =
      doc.querySelector('[itemprop="description"]')?.textContent?.trim() ||
      doc.querySelector("cite > span.hidden")?.textContent?.trim() ||
      null;
    const yearMatch = html.match(/Veröffentlicht:\s*(\d{4})/i);
    const year = yearMatch ? parseInt(yearMatch[1], 10) : null;
    const runtime =
      doc.querySelector("em") && /Spielzeit/i.test(html)
        ? (html.match(/Spielzeit:\s*<em>([^<]+)<\/em>/i) || [])[1]
        : null;

    const posterEl =
      doc.querySelector("img.cover2, img#img__" + (doc.querySelector("#viewID")?.getAttribute("data-id") || ""),) ||
      doc.querySelector('img[itemprop="image"], img[src*="/files/movies/"]');
    let poster = posterEl
      ? abs(pageUrl, posterEl.getAttribute("src") || posterEl.getAttribute("content") || "")
      : null;
    if (!poster) {
      const m = html.match(/\/files\/movies\/\d+\/[^"'\\\s>]+\.(?:jpg|jpeg|png|webp)/i);
      if (m) poster = abs(pageUrl, m[0]);
    }

    const genres = [];
    doc.querySelectorAll('#detail-content-list a[href*="/search/genre/"]').forEach((a) => {
      const g = cleanTitle(a.textContent);
      if (g && !genres.includes(g)) genres.push(g);
    });

    const hosters = parseHosters(html, pageUrl);
    const id = idHint || slugId(pageUrl, title);
    const movie = {
      id,
      title,
      overview,
      year,
      runtime,
      posterUrl: poster,
      backdropUrl: poster,
      detailPath: pageUrl,
      mediaKind: "movie",
      genres,
      hosters,
      seasons: [
        {
          number: 1,
          title: "Film",
          posterUrl: poster,
          backdropUrl: poster,
          episodes: [
            {
              id: `${id}-movie`,
              seriesId: id,
              seasonNumber: 1,
              number: 1,
              title: title || "Film",
              overview,
              stillUrl: poster,
              streamPageUrl: pageUrl,
              streamUrl: null,
            },
          ],
        },
      ],
    };
    return movie;
  }

  function parseHosters(html, pageUrl) {
    const doc = new DOMParser().parseFromString(html, "text/html");
    const hosters = [];
    doc.querySelectorAll("ul.currentStreamLinks").forEach((ul) => {
      const name = ul.querySelector(".hostName")?.textContent?.trim() || "Hoster";
      const a =
        ul.querySelector("a[data-player-url]") ||
        ul.querySelector("a.iconPlay[href], a.button[href]");
      if (!a) return;
      const raw = a.getAttribute("data-player-url") || a.getAttribute("href") || "";
      if (!raw || raw === "#") return;
      const url = abs(pageUrl, raw);
      const score =
        (window.SiteSearch?.scoreHosterName
          ? 0
          : 0) +
        (/voe/i.test(name) || /voe\.sx/i.test(url) ? 100 : 0) +
        (/vidara|vidnest/i.test(name) ? 70 : 0) +
        (/vidsonic/i.test(name) ? 40 : 0) +
        (/firestream/i.test(name) ? 30 : 0) +
        (/\bhd\b/i.test(name) ? 5 : 0);
      hosters.push({ provider: name, name, url, score, language: "" });
    });
    hosters.sort((a, b) => b.score - a.score);
    return hosters;
  }

  return {
    parseMovieList,
    parseMovieDetail,
    parseHosters,
    browsePaths,
    isMovieSite,
    isEpisodeLike,
    cleanTitle,
    slugId,
    abs,
  };
})();
